File: algorithms/cpp/17.电话号码的字母组合.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

/*
 * @lc app=leetcode.cn id=17 lang=cpp
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
class Solution {
 public:
  vector<string> letterCombinations(string digits) {
    vector<string> ans;
    dfs(digits, 0, "", ans);
    return ans;
  }

 private:
  unordered_map<char, string> hash{
      {'2', "abc"}, {'3', "def"},  {'4', "ghi"}, {'5', "jkl"},
      {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"},
  };
  void dfs(const string& digits, int i, const string& cur_ans,
           vector<string>& ans) {
    if (i >= digits.size() && !cur_ans.empty()) {
      ans.push_back(cur_ans);
      return;
    }
    string cur_str = hash[digits[i]];
    for (int j = 0; j < cur_str.size(); j++) {
      dfs(digits, i + 1, cur_ans + cur_str[j], ans);
    }
  }
};
// @lc code=end

File: algorithms/golang/17.电话号码的字母组合.go
/*
 * @lc app=leetcode.cn id=17 lang=golang
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start

/**
 * 回溯
 */
func letterCombinations(digits string) []string {
	ans := []string{}
	if len(digits) == 0 {
		return ans
	}
	m := []string{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}

	var recursion func(level int, cur string)
	recursion = func(level int, cur string) {
		if level == len(digits) {
			ans = append(ans, cur)
			return
		}
		for _, v := range m[digits[level]-'0'] {
			recursion(level+1, cur+string(v))
		}
	}
	recursion(0, "")
	return ans
}

// @lc code=end
